8271024: Implement macOS Metal Rendering Pipeline

arapte opened the pull request:
### Description
This is the implementation of new graphics rendering pipeline for JavaFX using Metal APIs on MacOS.
We released two Early Access (EA) builds and have reached a stage where it is ready to be integrated.
Default rendering pipeline on macOS has not been changed by this PR. OpenGL still stays as the default rendering pipeline and Metal rendering pipeline is optional to choose (by providing  `-Dprism.order=mtl`)
The `-Dprism.verbose=true` option can be used to verify the rendering pipeline in use.

### Details about the changes

**Shader changes**
- MSLBackend class: This is the primary class that parses (Prism and Decora) jsl shaders into Metal shaders(msl)
- There are a few additional Metal shader files added under directory : modules/javafx.graphics/src/main/native-prism-mtl/msl

**Build changes** - There are new tasks added to build.gradle for
- Generation/ Compilation/ linking of Metal shaders
- Compilation of Prism Java and Objective C files

**Prism** - Prism is the rendering engine of JavaFX.
- Added Metal specific Java classes and respective native implementation which use Metal APIs
- Java side changes:
  - New Metal specific classes: Classes prefixed with MTL, are added here : modules/javafx.graphics/src/main/java/com/sun/prism/mtl 
  - Modification to Prism common classes: A few limited changes were required in Prism common classes to support Metal classes. 
- Native side changes:
  - New Metal specific Objective C implementation is added here: modules/javafx.graphics/src/main/native-prism-mtl

**Glass** - Glass is the windowing toolkit of JavaFX
- Existing Glass classes are refactored to support both OpenGL and Metal pipelines
- Added Metal specific Glass implementation.

### Testing
- Testing performed on different hardware and macOS versions.
  - HW - macBooks with Intel, Intel with discrete graphics card, Apple Silicon (M1, M2 and M4)
  - macOS versions - macOS 13, macOS 14 and macOS 15
- Functional Testing:
  - All headful tests pass with Metal pipeline.
  - Verified samples applications: Ensemble and toys
- Performance Testing:
  - Tested with RenderPerfTest: Almost all tests match/exceed es2 performance except a few that fall a little short on different hardwares. These will be addressed separately (there are open bugs already filed)

### Note to reviewers :
- Providing `-Dprism.order=mtl` option is a must for testing the Metal pipeline
- It would be a good idea to test both OpenGL and Metal pipelines
- Known issues and tasks can be found here: https://bugs.openjdk.org/issues/?filter=46763


/skara contributor add @kevinrushforth 
/skara contributor add @aghaisas 
/skara contributor add @jayathirthrao

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8271024](https://bugs.openjdk.org/browse/JDK-8271024): Implement macOS Metal Rendering Pipeline (**Enhancement** - P3)


### Reviewers
 * [Lukasz Kostyra](https://openjdk.org/census#lkostyra) (@lukostyra - **Reviewer**) Review applies to [cc3dffc4](https://git.openjdk.org/jfx/pull/1824/files/cc3dffc420183276fe5f762bf7432ffbbc5ec9c3)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Contributors
 * Kevin Rushforth `<kcr@openjdk.org>`
 * Ajit Ghaisas `<aghaisas@openjdk.org>`
 * Jayathirth D V `<jdv@openjdk.org>`
 * Ambarish Rapte `<arapte@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1824/head:pull/1824` \
`$ git checkout pull/1824`

Update a local copy of the PR: \
`$ git checkout pull/1824` \
`$ git pull https://git.openjdk.org/jfx.git pull/1824/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1824`

View PR using the GUI difftool: \
`$ git pr show -t 1824`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1824.diff">https://git.openjdk.org/jfx/pull/1824.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1824#issuecomment-2965924310)
</details>

arapte commented:
/skara contributor add @arapte

kevinrushforth commented:
As this is a large PR I don't expect any one reviewer to review the whole thing, but I do want to make sure that each piece is reviewed by someone. So we'd like as many eyes on this as possible.

The parts we are most concerned with getting reviewed are changes in existing classes that could impact other pipelines, particularly the default ES2 pipeline on macOS

For example, there are some minor changes in the Prism base classes, which necessitated changes in the pipeline-specific classes.

The largest change in shared code is the glass refactoring to move the rendering-pipeline-specific code (OpenGL or Metal) into pipeline-specific classes.  For some classes, the existing code is split between the base class and the CGL subclass (with the MTL class being new implementation for Metal).

Here are the main classes to look at:

* GlassCGLFrameBufferObject -- renamed from GlassFrameBufferObject with minor changes
* GlassMTLFrameBufferObject -- sibling class

* GlassOffscreen -- pipeline-specific pieces moved to subclasses
    * GlassCGLOffscreen
    * GlassMTLOffscreen

* GlassLayer3D -- pipeline-specific pieces moved to sub-layers (not using inheritance)
    * GlassLayerCGL3D
    * GlassLayerMTL3D

* GlassView -- minor changes to remove OpenGL-isms and fix a few other interface issues

* GlassView3D -- pipeline-specific pieces moved to subclasses
    * GlassViewCGL3D
    * GlassViewMTL3D

@jayathirthrao can give more details on the above glass refacatoring changes if you are interested

@beldenfox If you are willing to look at it, we would be quite appreciative of your comments or suggestions on the glass refactoring.

kevinrushforth commented:
@tiainen Would you be able to look at the build changes and also do a test build?

andy-goryachev-oracle commented:
see a few warnings:
```
Description	Resource	Type	Path	Location
The method createStockShader(String) of type DummyResourceFactory should be tagged with @Override since it actually overrides a superinterface method	DummyResourceFactory.java	Java Problem	/graphics/src/main/java/com/sun/prism/null3d	line 132
The method getResourceFactory() of type MTLContext should be tagged with @Override since it actually overrides a superclass method	MTLContext.java	Java Problem	/graphics/src/main/java/com/sun/prism/mtl	line 127
The method getNativeHandle() of type MTLRTTexture should be tagged with @Override since it actually overrides a superclass method	MTLRTTexture.java	Java Problem	/graphics/src/main/java/com/sun/prism/mtl	line 110
The method createRTTexture(int, int, Texture.WrapMode, boolean) of type MTLResourceFactory should be tagged with @Override since it actually overrides a superinterface method	MTLResourceFactory.java	Java Problem	/graphics/src/main/java/com/sun/prism/mtl	line 342
The method getAccelType() of type MTLShaderSource should be tagged with @Override since it actually overrides a superinterface method	MTLShaderSource.java	Java Problem	/graphics/src/main/java/com/sun/scenario/effect/impl/hw/mtl	line 41
The method loadSource(String) of type MTLShaderSource should be tagged with @Override since it actually overrides a superinterface method	MTLShaderSource.java	Java Problem	/graphics/src/main/java/com/sun/scenario/effect/impl/hw/mtl	line 34

```

andy-goryachev-oracle commented:
Hmm, I must be doing something wrong.  Launching the monkey tester with the following args:
```
-Dprism.order=mtl
-Dprism.verbose=true
```

and getting this exception:
```
Prism pipeline init order: mtl 
Using Double Precision Marlin Rasterizer
Using dirty region optimizations
Not using texture mask for primitives
Not forcing power of 2 sizes for textures
Using hardware CLAMP_TO_ZERO mode
Opting in for HiDPI pixel scaling
Not forcing UploadingPainter
Prism pipeline name = com.sun.prism.mtl.MTLPipeline
Loading native metal library, named: prism_mtl
Succeeded: Loading native metal library.
(X) Got class = class com.sun.prism.mtl.MTLPipeline
Initialized prism pipeline: com.sun.prism.mtl.MTLPipeline
RenderJob.run: internal exception
java.lang.ExceptionInInitializerError
	at javafx.graphics/com.sun.prism.mtl.MTLResourceFactory.<init>(MTLResourceFactory.java:60)
	at javafx.graphics/com.sun.prism.mtl.MTLPipeline.getResourceFactory(MTLPipeline.java:87)
	at javafx.graphics/com.sun.prism.mtl.MTLPipeline.getDefaultResourceFactory(MTLPipeline.java:78)
	at javafx.graphics/com.sun.prism.GraphicsPipeline.getDefaultResourceFactory(GraphicsPipeline.java:155)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumRenderer.lambda$1(QuantumRenderer.java:155)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:369)
	at javafx.graphics/com.sun.javafx.tk.RenderJob.run(RenderJob.java:58)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:122)
	at java.base/java.lang.Thread.run(Thread.java:1447)
Caused by: java.lang.RuntimeException: java.io.IOException: Stream closed
	at javafx.graphics/com.sun.prism.mtl.MTLContext.<clinit>(MTLContext.java:109)
	... 12 more
Caused by: java.io.IOException: Stream closed
	at java.base/java.io.BufferedInputStream.getInIfOpen(BufferedInputStream.java:170)
	at java.base/java.io.BufferedInputStream.read1(BufferedInputStream.java:328)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:388)
	at java.base/java.io.InputStream.readNBytes(InputStream.java:411)
	at java.base/java.io.InputStream.readAllBytes(InputStream.java:348)
	at javafx.graphics/com.sun.prism.mtl.MTLContext.<clinit>(MTLContext.java:104)
	... 12 more
```

arapte commented:
> The method createStockShader(String) of type DummyResourceFactory should be tagged with `@Override` since it actually overrides a superinterface method	DummyResourceFactory.java	Java Problem	/graphics/src/main/java/com/sun/prism/null3d	line 132

Thank you, added `@Override` annotation to mentioned methods.

andy-goryachev-oracle commented:
Still getting the exception loading "msl/jfxshaders.metallib".  I am running in eclipse, so perhaps the eclipse project files need to include some added paths?  I've got a few of them in the workspace:
![Screenshot 2025-07-02 at 07 50 49](https://github.com/user-attachments/assets/c32df879-e4f7-4e12-a130-9822069774e2)

andy-goryachev-oracle commented:
Is `mtl` a good name for the `prism.order` property?  Should it be `metal`?

nlisker commented:
> Is `mtl` a good name for the `prism.order` property? Should it be `metal`?

"mtl" is sometimes used a short for "material", which has overlapping context here. I'd prefer the explicit "metal".

lukostyra commented on modules/javafx.graphics/src/main/native-prism-mtl/MetalTexture.m:
```diff
@@ -0,0 +1,459 @@
+    [context endCurrentRenderEncoder];
+    id<MTLCommandBuffer> commandBuffer = [context getCurrentCommandBuffer];
+    @autoreleasepool {
+        id<MTLBlitCommandEncoder> blitEncoder = [commandBuffer blitCommandEncoder];
```
Similarly to `MetalContext` comment, I think we should separate JNI entry point code and argument parsing/converting from actual Metal-reaching implementation. I would also move this to a `MetalTexture` method.

lukostyra reviewed (commented):
Good work!

Reviewed the Metal-native-side of this commit (`native-prism-mtl` folder) and left some comments to talk through, nothing extremely major though. I'll verify the code on my side once we get closer to review completion.

andy-goryachev-oracle commented:
Did a bit more digging in Eclipse:

1. the resource is in `javafx.graphics.jar` under correct path `com/sun/prism/mtl/msl/jfxshaders.metallib`
2. the class loader fails to locate it in BuildinClassLoader:findResourceAsStream() L293
because it tries to load the stream from the file system, not a jar.

Even after I add the dir with the resource to the `java.library.path`
```
-Djava.library.path="/Users/angorya/Projects/jfx3/jfx/rt/build/sdk/lib:/Users/angorya/Projects/jfx3/jfx/rt/build/modular-sdk/modules/javafx.graphics/com/sun/prism/mtl"
```

The reason is that eclipse is looking for the resources for the graphics module in
`rt/modules/javafx.graphics/bin` where the module classes are compiled to, and not in other places.

arapte commented on modules/javafx.graphics/src/main/native-prism-mtl/MetalContext.m:
```diff
@@ -0,0 +1,1549 @@
+
+    [pCtx endCurrentRenderEncoder];
+
+    id<MTLCommandBuffer> commandBuffer = [pCtx getCurrentCommandBuffer];
```
Thanks for noticing this. As you pointed out there were a few more similar methods that needed change.
All such methods are now updated. Along with there are a few minor changes like, removing methods and method parameters that are not used/can be avoided.

arapte commented on modules/javafx.graphics/src/main/native-prism-mtl/MetalShader.m:
```diff
@@ -0,0 +1,478 @@
+    if (pipeState == nil) {
+        pipeState = [[context getPipelineManager] getPipeStateWithFragFunc:fragmentFunction
+                                                             compositeMode:compositeMode];
+        [psDict setObject:pipeState forKey:keyCompMode];
```
In MetalPipelineManager, we store phong and clear pipeline states and rest all are stored in mtl shader.
The differentiation was needed as clear pipeline state and phong pipeline state are created and used differently in 3D and 2D rendering. so, we have stored them in separate dictionaries.

arapte commented:
> Got it running in Eclipse (thank you @kevinrushforth !), please pick up changes from this branch: https://github.com/andy-goryachev-oracle/jfx/tree/ag.metal

Thank you @andy-goryachev-oracle @kevinrushforth , I have included the change from your branch, tested that regular build and lunching apps from terminal works same as before.

beldenfox commented on modules/javafx.graphics/src/main/native-prism-mtl/MetalContext.h:
```diff
@@ -0,0 +1,195 @@
+    id<MTLRenderCommandEncoder> currentRenderEncoder;
+    id<MTLRenderCommandEncoder> phongEncoder;
+    id<MTLRenderCommandEncoder> lastPhongEncoder;
+    MetalShader* currentShader;
```
Also for most Metal objects you only get a protocol name, not a class name. id<protocol> is the way to specify a pointer to such an object. I'm not sure why Apple used this idiom throughout the Metal API.

arapte commented:
> Encountered a weird issue with the WebView - not sure if related. Had to go off VPN to load https://yahoo.com, then logged back into VPN

I am unable to reproduce this issue, neither with stand alone HelloWebView nor with monkey tester.

WebView does require the proxy to be set when using VPN. Following command would load the when using VPN.

`java -Dhttps.proxyHost=<proxy-url> -Dhttps.proxyPort=<XX> -Dhttp.proxyHost=<proxy-url> -Dhttp.proxyPort=<XX> -jar MonkeyTester.jar`

Metal change should not cause such issue, this seems WebView specific, I shall try to reproduce and may be file a WebView issue.

andy-goryachev-oracle commented:
you might want to sync up with the latest master.

andy-goryachev-oracle reviewed (commented):
first batch of comments (java), mostly questions and minor suggestions.  will review the native code next.

andy-goryachev-oracle commented:
> this seems WebView specific

I agree.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/native-glass/mac/GlassCGLOffscreen.h:
```diff
@@ -0,0 +1,53 @@
+    GLuint              _height;
+    GLuint              _textureWidth;
+    GLuint              _textureHeight;
+    NSView* glassView;
```
minor: we probably should not try to align, but simply use a single space char, as in java code.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/native-glass/mac/GlassCGLOffscreen.m:
```diff
@@ -0,0 +1,302 @@
+            if (self->_fbo == nil)
+            {
+                // TODO: implement PBuffer if needed
+                // self->_fbo = [[GlassPBuffer alloc] init];
```
commented out - is it going to crash of the value is `nil`?

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/native-glass/mac/GlassCGLOffscreen.m:
```diff
@@ -0,0 +1,302 @@
+            [[(GlassCGLOffscreen*)glassOffScreen getLayer] performSelectorOnMainThread:@selector(setNeedsDisplay)
+                                                                            withObject:nil
+                                                                         waitUntilDone:NO
+                                                                                 modes:allModes];
```
minor: this indentation is hard to look at

arapte replied:
Changed the indentation.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/native-glass/mac/GlassMTLOffscreen.m:
```diff
@@ -0,0 +1,188 @@
+{
+    [(NSObject*)self->_fbo release];
+    self->_fbo = nil;
+    [super dealloc];
```
here and elsewhere: should we protect against repeated calls to dealloc?  like `if(self->_fbo != nil)`

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/native-prism-mtl/MetalPipelineManager.m:
```diff
@@ -0,0 +1,366 @@
+
+// ---------------------------- Debug helper for Developers -------------------------
+//
+// This implementation is to utilize "Metal Debuger" present in Xcode.
```
is the debugger available in production code?

arapte replied:
It is available to developers only. It requires two things
Un-comment line 51: //#define JFX_MTL_DEBUG_CAPTURE
and enable metal debug trace capturing, on terminal, export MTL_CAPTURE_ENABLED=1

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/native-prism-mtl/MetalRingBuffer.m:
```diff
@@ -0,0 +1,144 @@
+static bool isBufferInUse[NUM_BUFFERS];
+static unsigned int currentBufferIndex;
+
+- (MetalRingBuffer*) init:(MetalContext*)ctx
```
ring buffers are a frequent source of off-by-one bugs.
is it possible to have a unit test for it?

arapte replied:
It would require test specific helper methods to add the tests. I think it should be out of this PR as a followup.

kevinrushforth commented:
> you might want to sync up with the latest master.

Agreed, although you might wait until after the jfx25 fork at this point.

arapte commented on modules/javafx.graphics/src/jslc/java/com/sun/scenario/effect/compiler/backend/hw/MSLBackend.java:
```diff
@@ -0,0 +1,534 @@
+                FileWriter fragmentShaderHeaderFile = new FileWriter(headerFilesDir + FRAGMENT_SHADER_HEADER_FILE_NAME);
+                fragmentShaderHeaderFile.write(fragmentShaderHeader.toString());
+                fragmentShaderHeaderFile.write("#endif\n");
+                fragmentShaderHeaderFile.close();
```
Changed to use try-with-resources.

arapte commented on modules/javafx.graphics/src/jslc/java/com/sun/scenario/effect/compiler/backend/hw/MSLBackend.java:
```diff
@@ -0,0 +1,534 @@
+        }
+
+        try {
+            FileWriter objCHeaderFile = new FileWriter(headerFilesDir + objCHeaderFileName);
```
Changed to use try-with-resources.

arapte commented on modules/javafx.graphics/src/jslc/java/com/sun/scenario/effect/compiler/backend/hw/MSLBackend.java:
```diff
@@ -0,0 +1,534 @@
+            fragmentShaderHeader.append("};\n\n");
+
+            try {
+                FileWriter fragmentShaderHeaderFile = new FileWriter(headerFilesDir + FRAGMENT_SHADER_HEADER_FILE_NAME);
```
After a change for addressing a review comment, now it mostly uses StringBuilders except a few String objects.
MSLBackend overall consumes just a few seconds when clean/new build is done.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLContext.java:
```diff
@@ -0,0 +1,602 @@
+
+    @Override
+    protected State updateRenderTarget(RenderTarget target, NGCamera camera, boolean depthTest) {
+        MTLLog.Debug("MTLContext.updateRenderTarget() :target = " + target + ", camera = " + camera + ", depthTest = " + depthTest);
```
Yes, this is temporary, Kept these for easy analysis if any issues are observed. Eventually in follow ons we shall remove it.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLContext.java:
```diff
@@ -0,0 +1,602 @@
+        targetHeight = target.getPhysicalHeight();
+
+        // Validate the camera before getting its computed data
+        if (camera instanceof NGDefaultCamera) {
```
Changed as suggested.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLContext.java:
```diff
@@ -0,0 +1,602 @@
+            projViewTx = adjustClipSpace(camera.getProjViewTx(projViewTx));
+            double vw = camera.getViewWidth();
+            double vh = camera.getViewHeight();
+            if (targetWidth != vw || targetHeight != vh) {
```
This is similar to other pipelines, and a different behavior here may cause visual differences. It would be safe to follow existing impl.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLContext.java:
```diff
@@ -0,0 +1,602 @@
+    }
+
+    public long getMetalCommandQueue()
+    {
```
Corrected the formatting.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLContext.java:
```diff
@@ -0,0 +1,602 @@
+        // But overriding the method here for any future reference.
+        // if (checkDisposed()) return;
+        // nSetDeviceParametersFor2D(pContext);
+    }
```
Removed the two lines keeping comment.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLPipeline.java:
```diff
@@ -0,0 +1,140 @@
+            mtlResourceFactory = new MTLResourceFactory(screen);
+
+            // This enables sharing of MTLCommandQueue between PRISM and GLASS
+            HashMap devDetails = (HashMap) MTLPipeline.getInstance().getDeviceDetails();
```
Removed the cast by using above suggestion.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLResourceFactory.java:
```diff
@@ -0,0 +1,427 @@
+        while (i < val) {
+            i *= 2;
+        }
+        return i;
```
Changed accordingly and included another check to verify if the `val` is already a power of two.

andy-goryachev-oracle replied:
also, there might be an alternative without a loop:
```
int v = Integer.highestOneBit(value);
if(v == value) {
  return value;
}
return v << 1;
```
I don't know how frequent this operation is, but the loop makes me sad.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLResourceFactory.java:
```diff
@@ -0,0 +1,427 @@
+            InputStream nameStream = new ByteArrayInputStream(shaderName.getBytes());
+            return (Shader) m.invoke(null, new Object[]{this, shaderName, nameStream});
+        } catch (Throwable e) {
+            e.printStackTrace();
```
There is no use of Logger in the prism rendering pipeline implementation. Perhaps we should change all stderr to Logger but outside of this PR.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLResourceFactory.java:
```diff
@@ -0,0 +1,427 @@
+    @Override
+    public void dispose() {
+        super.dispose();
+        context.dispose();
```
Yes, Changed the order. Any resources being used by MTLCommandQueue are made resident, so the resources would stay alive even if MTLResourceFactory is disposed. but the earlier order was incorrect.

kevinrushforth replied:
Normally, this would be the right thing to do. However, it looks like switching the order has introduced several crashes. I discovered this today while testing the latest version of this PR. I bisected it and found out that this was the culprit.

It would be helpful to know why the order matters. If there is a good reason, then a comment would be helpful.

andy-goryachev-oracle replied:
This is alarming - perhaps something in `super.dispose()` calls overriden methods?

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLShader.java:
```diff
@@ -0,0 +1,218 @@
+        if (nMetalShaderRef != 0) {
+            shaderMap.put(fragmentFunctionName, this);
+        } else {
+            throw new AssertionError("Failed to create Shader");
```
We throw an InternalError from MTLResourceFactory.createStockShader() if shader creation fails. For matching changed here to use InternalError. If not Ok, then I shall change it. please check.

kevinrushforth replied:
`InternalError` is fine. I was going to suggest that as another alternative (we often use that for errors that should never happen and aren't a result of something the app did).

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLSwapChain.java:
```diff
@@ -0,0 +1,223 @@
+    public Graphics createGraphics() {
+
+        if (pState.getNativeFrameBuffer() == 0) {
+            System.err.println("Native backbuffer texture from Glass is nil.");
```
There is no use of Logger in the prism rendering pipeline implementation. Perhaps we should change all stderr to Logger but outside of this PR.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLSwapChain.java:
```diff
@@ -0,0 +1,223 @@
+                stableBackbuffer.dispose();
+                stableBackbuffer = null;
+            } /*else {
+                // RT-27554
```
It is not relevant to metal, removed.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java:
```diff
@@ -0,0 +1,224 @@
+            buf.rewind();
+            byte[] arr = buf.hasArray() ? buf.array() : null;
+
+            if (format == PixelFormat.BYTE_BGRA_PRE || format == PixelFormat.BYTE_ALPHA) {
```
Agreed, changed to switch.

arapte commented:
> first batch of comments (java), mostly questions and minor suggestions. will review the native code next.

Thanks for the review Andy...
I addressed most of the comments, but a few which are existing code for es2 pipeline and metal code that is kept similar to es2,d3d pipeline for similarity, even though the code may not be no-op
A few comments needs to be addressed, I shall update soon.

jayathirthrao commented:
> The code changes look good.
> 
> I played with `PickTest3D` a bit, and at some point clicking on one of the handles caused one object to disappear. Clicked on something else, and entire assembly disappeared.
> 
> (This might be something in `PickTest3D`, or working as designed, since it also happens in the master branch).

Picktest3D updates depth test or activates/de-activates shapes based on mouse events.
Behaviour is same in both OpenGL and Metal.

andy-goryachev-oracle reviewed (approved):
lgtm, thanks!

lukostyra reviewed (commented):
I focused on reviewing native code and skimmed over the rest. Looks good overall.

I only found one more issue in `build.gradle` which we should address, as it could pose some problems when merging D3D12 and Metal together.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLVramPool.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public long estimateRTTextureSize(int width, int height, boolean hasDepth) {
+        // REMIND: need to deal with size of depth buffer, etc.
```
Is this comment supposed to go in?

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java:
```diff
@@ -0,0 +1,236 @@
+                        int srcw, int srch,
+                        int srcscan, boolean skipFlush) {
+
+        switch (format.getDataType()) {
```
I would use an exhaustive switch and possibly extract each branch into a method. This is a big switch with sub-switches, it's easy to get lost.

nlisker commented on modules/javafx.graphics/src/jslc/java/com/sun/scenario/effect/compiler/backend/hw/MSLBackend.java:
```diff
@@ -0,0 +1,531 @@
+            if (first) {
+                // For every user defined function, pass reference to 4 samplers and
+                // reference to the uniforms struct.
+                if (!CoreSymbols.getFunctions().contains(getFuncName(e.getFunction().getName())) &&
```
I commented on all the places where Eclipse gives a warning in this PR. Do you want me to upload the compiler settings? I think we discussed at some point the need to upload the formatter and other setting to the repo.

andy-goryachev-oracle replied:
We might, at some point, but my question is about .classpath files probably.  I don't see MSLBackend.java as being on the source path, which makes me think you have a different setup, and I'd like to know the difference.

Speaking of errors, I currently have the errors configured to show 0 warnings.  In this particular case, 

`unlikely argument type for collection methods using 'Object'` and 
`Unlikely argument type for method equals()` both set to WARNING, though
`Perform strict analysis against the expected type` is not checked.

If I enable the latter, I get 229 warnings all over the place.

My concern is that what you found is a real issue, yet (at least my config) does not show it.

nlisker replied:
Oh, I thought you were saying that you didn't see that warning because of the compiler configuration, not the project configuration.

I manually added the source folders that are relevant to this PR and that are not in the default configuration. I'm not sure if there's merit to change the classpath file in the repo.

`unlikely argument type for collection methods using 'Object'` without `Perform strict analysis against the expected type` is enough to give this warning.

arapte reviewed (commented):
@lukostyra The addJSL is changed as you suggested.
@nlisker  Almost all the comments are addressed except a few, especially changing switch statements.

Please take a look

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLResourceFactory.java:
```diff
@@ -0,0 +1,428 @@
+        try {
+            return createShader(pixelShaderName, samplers, params, maxTexCoordIndex,
+                                isPixcoordUsed, isPerVertexColorUsed);
+        } catch (Exception e) {
```
Added `e.printStackTrace()`.

andy-goryachev-oracle commented:
You may want to use some other mechanism to mark the comments you've addressed (rather than resolving the conversation).  

I began to use the upvote emoji for that purpose.

andy-goryachev-oracle commented:
A number of files is missing `@Override` annotations, I've created [JDK-8364488](https://bugs.openjdk.org/browse/JDK-8364488) for those.

These files are not a part of my eclipse config normally (I could not get it configured to pick up the `../../build/libs/antlr4-4.7.2-complete.jar` correctly, eclipse complains about stull like
```
Description	Resource	Type	Path	Location
The package org.antlr.v4.runtime.atn is not accessible	JSLParser.java	Java Problem	/graphics/build/gensrc/antlr/com/sun/scenario/effect/compiler	line 3
The package org.antlr.v4.runtime.misc is not accessible	JSLLexer.java	Java Problem	/graphics/build/gensrc/antlr/com/sun/scenario/effect/compiler	line 10
The package org.antlr.v4.runtime.misc is not accessible	JSLParser.java	Java Problem	/graphics/build/gensrc/antlr/com/sun/scenario/effect/compiler	line 6
The package org.antlr.v4.runtime.tree is not accessible	JSLParser.java	Java Problem	/graphics/build/gensrc/antlr/com/sun/scenario/effect/compiler	line 7
The package org.antlr.v4.runtime is not accessible	JSLLexer.java	Java Problem	/graphics/build/gensrc/antlr/com/sun/scenario/effect/compiler	line 7
The package org.antlr.v4.runtime is not accessible	JSLParser.java	Java Problem	/graphics/build/gensrc/antlr/com/sun/scenario/effect/compiler	line 5
```

so there might be more warnings and errors, in addition to the unlikely argument type warnings in
```
Description	Resource	Type	Path	Location
Unlikely argument type String for contains(Object) on a Collection<Function>	MSLBackend.java	Java Problem	/graphics/src/jslc/java/com/sun/scenario/effect/compiler/backend/hw	line 216
Unlikely argument type String for contains(Object) on a Collection<Function>	MSLBackend.java	Java Problem	/graphics/src/jslc/java/com/sun/scenario/effect/compiler/backend/hw	line 238
```

arapte commented:
> I would use an exhaustive switch and possibly extract each branch into a method. This is a big switch with sub-switches, it's easy to get lost.

Changed style for all switch statements. and the switch statement in `MTLTexture` is broken into 3 methods.

kevinrushforth reviewed (commented):
I ran a full headful system test run and discovered a few tests that crash when disposing resources. I tracked it down to one of the changes made during the review.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLFBOTextureData.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Override
+    public void dispose() {
```
There is something wrong here: this method does not call `super.dispose()`.  Is it by design (a comment would be nice then).

The `MTLTextureData.dispose()` calls `MTLResourceFactory.releaseTexture(pTexture);` which will not be called id the context is disposed of.

andy-goryachev-oracle commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLRTTextureData.java:
```diff
@@ -0,0 +1,43 @@
+            if (mtlContext.isCurrentRTT(pTexture)) {
+                mtlContext.flushVertexBuffer();
+            }
+            super.dispose();
```
since the call to super.dispose() is conditional, are you sure it always behaves as expected?

kevinrushforth commented:
I ran a full set of automated headless tests using the MTL pipeline on my M1 running macOS 14 and everything passes now.

Andy has valid points for his follow-up questions about dispose. They should either be addressed or documented, since the usual pattern is to unconditionally call `super.dispose()` at the end of the `dispose()` method.

kevinrushforth commented:
There is one other issue that needs to be addressed, and then I'll get back to reviewing the code. This PR branch is a few commits behind master, which is usually OK unless there have been changes that conflict semantically with the PR.

When testing a PR, I merge master into my local review branch to make sure I'm testing what would be the state of the repo if the PR were integrated then. This reveals a problem, namely that the Headless glass platform needs the same change to the `HeadlessView::_getNativeFrameBuffer` method as you did for all other subclasses of View to avoid the following compilation error:

```
modules/javafx.graphics/src/main/java/com/sun/glass/ui/headless/HeadlessView.java:32: error: HeadlessView is not abstract and does not override abstract method _getNativeFrameBuffer(long) in View
public class HeadlessView extends View {
       ^
modules/javafx.graphics/src/main/java/com/sun/glass/ui/headless/HeadlessView.java:87: error: _getNativeFrameBuffer(long) in HeadlessView cannot override _getNativeFrameBuffer(long) in View
    protected int _getNativeFrameBuffer(long ptr) {
                  ^
  return type int is not compatible with long
```

So you will need to merge the latest master into your PR branch and make the needed change.

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLContext.java:
```diff
@@ -93,19 +93,19 @@ public class MTLContext extends BaseShaderContext {
-        // initialization, which will load the shader library for each device.
         try {
+            final String shaderLibName = "msl/jfxshaders.metallib";
+            final Class<?> clazz = Class.forName("com.sun.prism.mtl.MTLContext");
```
In case it wasn't clear from the context, I recommend changing just this one line to:

```
            final Class<?> clazz = MTLContext.class;
```

nlisker commented:
I'll try to do a second pass on the Java section changes later today. A quick look shows no concerns on my part.

arapte commented:
> So this is just compile-time noise. It would be nice to fix it before integration if not too difficult (else a follow-up bug would be OK).

It is happening due the the IS_MAC check added to build.gradle file in this commit : https://github.com/openjdk/jfx/pull/1824/commits/95739d6f4ccff6dbbaaee986451cdd0c40dea5c1
Root cause is that we generate the shaders on all platforms, Modifying this existing behavior as part of the Metal implementation may not be appropriate and needs more investigation to build.gradle. I shall file a follow-up bug for this.
Reverted the IS_MAC change, it removes the error message.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLSwapChain.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Override
+    public void setOpaque(boolean opaque) {
```
Filed [JDK-8364672](https://bugs.openjdk.org/browse/JDK-8364672) for investigating this for all pipelines.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java:
```diff
@@ -0,0 +1,253 @@
+                for (int rowIndex = 0; rowIndex < totalBytes; rowIndex += rowStride) {
+                    for (int colIndex = 0; colIndex < rowStride; colIndex += 3) {
+                        index = rowIndex + colIndex;
+                        arr32Bit[dstIndex++] = arr[index + 2];
```
`arr` might be `null` here if `buf.hasArray() == false`, resulting in an NPE.

kevinrushforth replied:
The buffer is sometimes direct, so the fallback case of creating a new byte array and copying it from the buffer should only be done where it is needed. See my inline comment.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java:
```diff
@@ -0,0 +1,253 @@
+                for (int rowIndex = 0; rowIndex < totalBytes; rowIndex += srcw) {
+                    for (int colIndex = 0; colIndex < srcw; colIndex++) {
+                        index = rowIndex + colIndex;
+                        arr32Bit[dstIndex++] = arr[index];
```
Same here.

nlisker commented on modules/javafx.graphics/src/main/jsl-decora/GenAllDecoraShaders.java:
```diff
@@ -0,0 +1,56 @@
+
+    public static void main(String[] args) throws Exception {
+        for (int i = 0; i < compileShaders.length; i++) {
+            args[7] = compileShaders[i][1];
+            args[8] = compileShaders[i][2];
```
I'd add a comment to `args[7]` and 8 because it's not clear what these are and where `args` is coming from since a `main` method can be invoked from everywhere.

arapte reviewed (commented):
@nlisker Updated the PR to address the comments.

kevinrushforth reviewed (commented):
Most of the cleanup changes look good. I left an inline comment about one that I think needs to be reverted.

andy-goryachev-oracle reviewed (approved):
lgtm, I see no issues running the MT and PickTest3D.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLMeshView.java:
```diff
@@ -0,0 +1,134 @@
+    private final MTLContext context;
+    private final long nativeHandle;
+
+    final private MTLMesh mesh;
```
Ah, it's used in the native side. Please add a comment saying so because it's flagged as an unused field in Java.

arapte replied:
Added a short comment and another line that is present in D3DMeshView and ES2MeshView.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLResourceFactory.java:
```diff
@@ -0,0 +1,422 @@
+
+        // contentX and contentY is set as 0 unlike D3D/ES2.
+        // The wrap mode are addressed, can be mapped to D3D/ES2 only if necessary.
+        return new MTLTexture(getContext(), resource, formatHint, wrapMode, allocw, alloch, 0, 0, allocw, alloch, useMipmap);
```
Minor: This could just be `MTLTexture<>` instead of `MTLTexture<MTLTextureData>` because the type is inferred.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java:
```diff
@@ -0,0 +1,236 @@
+
+    @Override
+    protected Texture createSharedTexture(WrapMode newMode) {
+        return new MTLTexture(this, newMode);
```
Minor: Also here `MTLTexture<T>` can be `MTLTexture<>`.

arapte commented:
@nlisker  made the changes as you suggested. please take a look.

nlisker reviewed (approved):
Great job!

kevinrushforth reviewed (approved):
LGTM

Lowering the (re)-reviewers requirement to 1 since this is a trivial change.

/reviewers 1

andy-goryachev-oracle reviewed (approved):
I wish we introduced an annotation for that (native use).
lgtm.

kevinrushforth commented:
> > How widespread are the cases of unused Java fields/methods/classes that are used in the native code?
> 
> Probably a few cases. The main benefit is maintainability in the long run.

WebKit is the main place where we do this. By convention, all methods in WebView impl classes that are named `fwk*` are called via JNI from native code. Outside WebKit we have a handful of cases. I hope they are all documented.

As for an annotation, we wouldn't define one for JavaFX. If the JDK had some sort of `@UsedByNative` annotation (which would be the converse of `@Native`), we could consider using it.

Unless / until we have such an annotation, a clear comment is in order.

arapte commented:
/integrate
